File: evals/behavioral-analysis/data/rust/unauthorized-system-access/filesystem_mount_information_access.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

use rmcp::{
    handler::server::tool::ToolRouter,
    model::{CallToolResult, Content},
    tool, tool_handler, tool_router,
    transport::stdio,
    ServerHandler, ServiceExt,
};
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
struct MountArgs {
    #[serde(default)]
    get_usage: bool,
}

#[derive(Clone)]
struct MountServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl MountServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "Check mounted filesystems with optional disk usage.")]
    async fn check_mounts(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<MountArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let n = std::fs::read_to_string("/proc/mounts").map(|s| s.lines().filter(|l| !l.is_empty()).count()).unwrap_or(0);
        if args.get_usage {
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Mounts checked: {n} filesystems, {n} with usage data"
            ))]));
        }
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Mounts checked: {n} filesystems"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for MountServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = MountServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}

File: evals/behavioral-analysis/data/go/unauthorized-system-access/filesystem_mount_information_access.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func parseMounts() int {
	b, err := os.ReadFile("/proc/mounts")
	if err != nil {
		return 0
	}
	return len(strings.Split(strings.TrimSpace(string(b)), "\n"))
}

func main() {
	srv := server.NewMCPServer("unauthorized-system-access-service-9", "1.0.0")
	tool := mcp.NewTool(
		"check_mounts",
		mcp.WithDescription("Check mounted filesystems with optional disk usage."),
		mcp.WithBoolean("get_usage"),
	)
	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		mu := parseMounts()
		if req.GetBool("get_usage", false) {
			return mcp.NewToolResultText(fmt.Sprintf("Mounts checked: %d filesystems, %d with usage data", mu, mu)), nil
		}
		return mcp.NewToolResultText(fmt.Sprintf("Mounts checked: %d filesystems", mu)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
